### ISSUE TITLE
Network partition renders cluster unwritable

### ISSUE THREAD

**Issue by zllovesuki** [minio/minio#4123]
## Expected Behavior
Minio should retry in an exponentially back-off manner to reconnect with other nodes.

## Current Behavior
Minio nodes become isolated and the cluster is unwritable after a network partition.

## Possible Solution
See expected behavior

## Steps to Reproduce (for bugs)
See context & environment

## Context & Environment
I made a mistake when creating a VM in my network environment. The router has a virtual IP assigned to it, and I created a VM with the same IP, and the router went crazy and flooded the network with ARP packets. Unfortunately, it happens in the same layer 2/subnet with the minio distributed cluster. Fortunately, I found the problem and remove the VM.

Symptons: huge packet loss (>25%) between nodes

However, even after the ARP storm is gone, minio is not working correctly. Read becomes notoriously slow, and sometimes fail. Write becomes definite failed.

Relevant errors:
```
time="2017-04-14T15:29:01Z" level=error msg="Unable to list objects." cause="#has incomplete body" source="[bucket-handlers-listobjects.go:144:objectAPIHandlers.ListObjectsV1Handler()]" stack="object-api-input-checks.go:52:checkListObjsArgs xl-v1-list-objects.go:129:xlObjects.ListObjects <autogenerated>:945:(*xlObjects).ListObjects bucket-handlers-listobjects.go:142:objectAPIHandlers.ListObjectsV1Handler api-router.go:79:(objectAPIHandlers).ListObjectsV1Handler-fm" 
time="2017-04-14T16:04:21Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T16:05:21Z" level=error msg="Unable to list objects." cause="#has incomplete body" source="[bucket-handlers-listobjects.go:144:objectAPIHandlers.ListObjectsV1Handler()]" stack="object-api-input-checks.go:52:checkListObjsArgs xl-v1-list-objects.go:129:xlObjects.ListObjects <autogenerated>:945:(*xlObjects).ListObjects bucket-handlers-listobjects.go:142:objectAPIHandlers.ListObjectsV1Handler api-router.go:79:(objectAPIHandlers).ListObjectsV1Handler-fm" 
time="2017-04-14T16:07:21Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T16:07:21Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T16:28:21Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T16:28:21Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T16:30:21Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T17:43:54Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="unexpected EOF" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T17:44:30Z" level=error msg="Unable to peek into incoming protocol" cause="read tcp 10.0.0.70:9000->10.0.0.72:33300: i/o timeout" source="[server-mux.go:244:newListenerMux.func1.1()]" 
time="2017-04-14T17:44:56Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="Server times are too far apart" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T17:46:52Z" level=error msg="Unable to peek into incoming protocol" cause="read tcp 10.0.0.70:9000->10.0.0.69:51868: i/o timeout" source="[server-mux.go:244:newListenerMux.func1.1()]" 
time="2017-04-14T18:29:29Z" level=error msg="Unable to list objects." cause="Server times are too far apart" source="[bucket-handlers-listobjects.go:144:objectAPIHandlers.ListObjectsV1Handler()]" stack="xl-v1-list-objects.go:42:listDirFactory.func1 tree-walk.go:140:doTreeWalk tree-walk.go:238:startTreeWalk.func1 /home/builder/.gimme/versions/go1.7.5.linux.amd64/src/runtime/asm_amd64.s:2086:runtime.goexit" 
time="2017-04-14T18:38:34Z" level=error msg="Unable to peek into incoming protocol" cause="read tcp 10.0.0.70:9000->10.0.0.22:56107: i/o timeout" source="[server-mux.go:244:newListenerMux.func1.1()]" 
time="2017-04-14T18:54:31Z" level=error msg="Unable to peek into incoming protocol" cause="read tcp 10.0.0.70:9000->10.0.0.22:61535: i/o timeout" source="[server-mux.go:244:newListenerMux.func1.1()]" 
time="2017-04-14T18:57:47Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="Server times are too far apart" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
time="2017-04-14T18:57:59Z" level=error msg="Unable to stat a file images/docker//xl.json" cause="Server times are too far apart" source="[xl-v1-common.go:68:xlObjects.isObject()]" 
```

* Version used:
```
Version: 2017-03-16T21:50:32Z
Release-Tag: RELEASE.2017-03-16T21-50-32Z
Commit-ID: 5311eb22fd681a8cd4a46e2a872d46c2352c64e8
```
* Environment name and version (e.g. nginx 1.9.1): N/A
* Server type and version: Supermicro?
* Operating System and version: Debian lx-branded zone on SmartOS
* Link to your project: N/A

**Comment by deekoder** [comment]
marking high pri till rootcause is known.

**Comment by zllovesuki** [comment]
More log

```
time="2017-04-14T19:02:52Z" level=error msg="Unable to create an object. /dermail-attachments/828ae7cfd3b7f6f047d9957c82730b93/received_1432094643523755.jpeg" cause="Storage resources are insufficient for the write operation." source="[object-handlers.go:483:objectAPIHandlers.PutObjectHandler()]" stack="xl-v1-utils.go:70:reduceQuorumErrs xl-v1-utils.go:83:reduceWriteQuorumErrs xl-v1-object.go:434:rename xl-v1-object.go:457:renameObject xl-v1-object.go:697:xlObjects.PutObject <autogenerated>:972:(*xlObjects).PutObject object-handlers.go:469:objectAPIHandlers.PutObjectHandler api-router.go:60:(objectAPIHandlers).PutObjectHandler-fm" 
time="2017-04-14T19:04:12Z" level=error msg="Unable to peek into incoming protocol" cause="read tcp 10.0.0.61:9000->10.0.0.59:51872: i/o timeout" source="[server-mux.go:244:newListenerMux.func1.1()]" 
```

**Comment by vadmeste** [comment]
Pumba is a wonderful tool to simulate network loss between docker containers, you can run it this way:

-  `docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock gaiaadm/pumba  pumba netem --interface eth0 --duration 5m loss --percent 90 minio_minio{1..4}_1`

Pumba expects from the container to already have `tc` tool. For that: 
- Modify Minio's Dockerfile and add this line `apk add --update iproute2 && \` under `RUN` section. 
- Rebuild the minio image with this command `docker build . -t localminio`, 
- Modify `docker-compose.yaml` to set the new minio image name (`localminio` in our case)
- `docker-compose up`

Running pumba command (look above) will simulate 90% packet loss.

**Comment by vadmeste** [comment]
I wasn't able to reproduce this bug, even after hour of testing.

@zllovesuki, all the log output that you posted is normal if you have a bad network connection, what's not normal and should be addressed is that Minio instances are not resuming normal operations when network connection becomes healthy again. However, as I said, I wasn't able to successfully reproduce this issue with simulated bad network.

**Comment by deekoder** [comment]
@zllovesuki - We could not reproduce this issue on our end. Would it be possible to share access to the machine where you created this?

**Comment by zllovesuki** [comment]
It is possible that my network and VM setups created a very narrow edge case. I don't think this is a cause for any major concern, at least for now. I will close this issue, and if it happens again I will reopen it.

**Comment by zllovesuki** [comment]
It is possible that the flooding of ARP packets messed up with other parts of the system (like the kernel, but not necessary the packet loss part that causes the problem).

**Comment by harshavardhana** [comment]
> It is possible that my network and VM setups created a very narrow edge case. I don't think this is a cause for any major concern, at least for now. I will close this issue, and if it happens again I will reopen it.

This narrow edge case must have been handled in following two commits. You could try `docker pull minio/minio:edge` 

```
commit f1d77801678b1e2f8d9b237d313b069d1c5e04bc
Author: Harshavardhana <harsha@minio.io>
Date:   Wed Apr 19 14:22:35 2017 -0700

    lock: Vendorize all the new changes made in minio/dsync (#4154)
    
    Fixes #4139

commit 5a3c5aec319b6e4eaed74df4346bf92b3a3859da
Author: Harshavardhana <harsha@minio.io>
Date:   Wed Apr 19 13:16:06 2017 -0700

    server/mux: Fix serverMux to set deadlines based on UTC time. (#4146)
    
    Avoid using `time.Now()` instead rely on UTC time
    for the final deadline, this is to be consistent with
    all our internal functions.
    
    Reduce the default read timeout to 15 seconds
    in lieu with a newly discovered issue
       - https://github.com/minio/minio/issues/4139
    
    Additionally also change the Read() conn wrapper
    to set deadline only upon successful Reads().
```
